### PR TITLE
Event details

### PR DESCRIPTION
**Pull request by fourlastor**
## Problem

We need event details for our apps!

## Solution

Some stuff has changed, I had to hack around a solution for the inconsistent data we have in the db. Long term solution is to validate this data before it gets written.

### CODE REVIEW

File: functions/src/firestore/data.ts [hunk 1]
```diff
@@ -24,11 +26,11 @@ export interface UserData {
 export interface EventData {
     readonly day: Reference<DayData>
     readonly end_time: Date
-    readonly place: Reference<PlaceData>
+    readonly place: Reference<PlaceData> | null
```
Comment (rock3r) [inline functions/src/firestore/data.ts]: Why not optional here?

File: functions/src/firestore/data.ts [hunk 2]
```diff
@@ -24,11 +26,11 @@ export interface UserData {
...
     readonly start_time: Date
     readonly submission: Reference<SubmissionData>
-    readonly track: Reference<TrackData>
-    readonly type: string
+    readonly track: Reference<TrackData> | null
```
Comment (rock3r) [inline functions/src/firestore/data.ts]: Ditto

File: functions/src/optional.ts [hunk 3]
```diff
@@ -3,3 +3,13 @@ export type Optional<T> = T | null | undefined
 export function present<T>(optional: Optional<T>): optional is T {
     return optional !== null && optional !== undefined
 }
+
+const identity = (a: any) => a
+
+export function map<T, R>(it: T | null | undefined, fn: (t: T) => R = identity): R | null {
+    if (!it) {
```
Comment (rock3r) [inline functions/src/optional.ts]: Maybe transform to positive? `if (it) { return fn(it) } else { return null }`

File: functions/src/firestore/data.ts (hunk 1 above)
Comment (fourlastor) [reply functions/src/firestore/data.ts]: Didn't find a case in which references were undefined, only null, but can do

File: functions/src/optional.ts (hunk 3 above)
Comment (fourlastor) [reply functions/src/optional.ts]: Yeah makes sense, will do

File: functions/src/firestore/data.ts (hunk 1 above)
Comment (rock3r) [reply functions/src/firestore/data.ts]: Would be mostly for consistency but not overly fussed

File: functions/src/optional.ts (hunk 3 above)
Comment (fourlastor) [reply functions/src/optional.ts]: db9b441

File: functions/src/firestore/data.ts (hunk 1 above)
Comment (fourlastor) [reply functions/src/firestore/data.ts]: 549206f

**Pull request merged**